### PR TITLE
feat: Initial API surface validator

### PR DESCRIPTION
**Pull request by ebariche**
## PR Type

What kind of change does this PR introduce?
- Feature

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

Internal Issue (If applicable):
unoplatform/private#219

### CODE REVIEW

File: src/Uno.APISurfaceValidator/Program.cs [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+			var filePath = UnpackArchive(args[0]);
+
+			Console.WriteLine($"Validating package {args[0]}");
+
+			foreach (var contract in _contractsToValidate)
+			{
+				Console.WriteLine($"Validating {Path.GetFileName(contract.path)} :");
+
+				var referenceAssembly = ReadAssemblyDefinition(contract.path);
+				var assembly = ReadAssemblyDefinition(Path.Combine(filePath, "lib", "netstandard2.0", contract.targetAssembly));
```
Comment (davidjohnoliver) [inline src/Uno.APISurfaceValidator/Program.cs:38]: What about Android, iOS, macOS?

File: src/Uno.APISurfaceValidator/Program.cs [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+			Func<TypeDefinition, bool> predicate = pattern[0] switch
+			{
+				'+' => t => t.IsPublic && t.FullName.StartsWith(pattern[1..]),
+				'-' => t => t.IsPublic && !t.FullName.StartsWith(pattern[1..]),
+				_   => t => t.IsPublic
+			};
+
+			foreach(var referenceType in referenceTypes.Where(predicate))
+			{
+				if(referenceType.FullName == "Windows.UI.Xaml.Documents.TextElement")
```
Comment (davidjohnoliver) [inline src/Uno.APISurfaceValidator/Program.cs:62]: Why a special case for `TextElement`?

File: src/Uno.APISurfaceValidator/Program.cs (hunk 1 above)
Comment (jeromelaban) [reply src/Uno.APISurfaceValidator/Program.cs:38]: Not yet, will be added in subsequent phases

File: src/Uno.APISurfaceValidator/Program.cs (hunk 2 above)
Comment (ebariche) [reply src/Uno.APISurfaceValidator/Program.cs:62]: IDK, I asked magic-8ball and it said "Yes" !

image

I mean, who am I to question the ball ?

Comment (ebariche) [reply src/Uno.APISurfaceValidator/Program.cs]: It's fixed :)

**Review by davidjohnoliver — APPROVED** [review]

**Review by jeromelaban — APPROVED** [review]

File: src/Uno.APISurfaceValidator/Program.cs [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+		{
+			(@"C:\Program Files (x86)\Windows Kits\10\References\10.0.19041.0\Windows.Foundation.FoundationContract\4.0.0.0\Windows.Foundation.FoundationContract.winmd", "Uno.Foundation.dll", "*"),
+			(@"C:\Program Files (x86)\Windows Kits\10\References\10.0.19041.0\Windows.Foundation.UniversalApiContract\10.0.0.0\Windows.Foundation.UniversalApiContract.winmd", "Uno.dll", "-Windows.UI.Xaml"),
+			(@"C:\Program Files (x86)\Windows Kits\10\References\10.0.19041.0\Windows.Phone.PhoneContract\1.0.0.0\Windows.Phone.PhoneContract.winmd", "Uno.dll", "*"),
+			(@"C:\Program Files (x86)\Windows Kits\10\References\10.0.19041.0\Windows.Networking.Connectivity.WwanContract\2.0.0.0\Windows.Networking.Connectivity.WwanContract.winmd", "Uno.dll", "*"),
+			(@"C:\Program Files (x86)\Windows Kits\10\References\10.0.19041.0\Windows.ApplicationModel.Calls.CallsPhoneContract\5.0.0.0\Windows.ApplicationModel.Calls.CallsPhoneContract.winmd", "Uno.dll", "*"),
+			(@"C:\Program Files (x86)\Windows Kits\10\References\10.0.19041.0\Windows.Foundation.UniversalApiContract\10.0.0.0\Windows.Foundation.UniversalApiContract.winmd", "Uno.UI.dll", "+Windows.UI.Xaml")
```
Comment (jeromelaban) [inline src/Uno.APISurfaceValidator/Program.cs:21]: We'll need to find a way to sync with the Reference project

**Review by jeromelaban — APPROVED** [review]

**Pull request merged**